File: argon-plugin-tube/shared/src/main/scala/dev/argon/plugin/tube/TubeBackend.scala
package dev.argon.plugin.tube

import dev.argon.options.*
import dev.argon.plugin.*
import dev.argon.util.*
import dev.argon.io.*
import zio.*

object TubeBackend extends Backend[Nothing] {
  override type OutputOptionID = TubeOutputOptionID
  override val outputOptions: OptionsHandler[OutputOptionID, [_] =>> SingleFile] = new OptionsHandlerImpl[OutputOptionID, [_] =>> SingleFile]
  override val outputResourceTags: Options[ResourceTag, OutputOptionID] = Options.fromFunction([E] => (id: OutputOptionID with TypedOptionID[E]) =>
    id match {
      case TubeOutputOptionID.ImplementationTube => TubeImplementationResourceTag
    } : ResourceTag[E]
  )

  override def emitModule(options: Options[Id, Nothing])(tube: SerializedTube): Options[Id, OutputOptionID] = ???
}

sealed trait TubeOutputOptionID extends OptionID derives CanEqual {
  override type ElementType <: Resource
  override type Decoded[_] = SingleFile
}

object TubeOutputOptionID {
  case object ImplementationTube extends OptionIDBase[[_] =>> SingleFile, TubeImplementationResource] with TubeOutputOptionID:
    override val info: OptionInfo[TubeImplementationResource] =
      OptionInfo(
        name = "tube.implementation",
        description = "Tube contaning implementation of all methods",
      )
}

File: plugins/java/tube/src/dev/argon/plugins/tube/TubeBackend.java
package dev.argon.plugins.tube;

import dev.argon.plugin.api.*;
import dev.argon.plugin.api.tube.*;
import java.util.Set;

final class TubeBackend<E extends Exception> implements Backend<E, TubeOptions<E>, TubeOutput<E>> {
    @Override
    public TubeOutput<E> emitModule(TubeOptions<E> options, Set<Platform<E, ?, ?, ?>> platforms, SerializedTube<E> tube) {
        return new TubeOutput<E>(
            new SerializedTubeResource<E>(tube, platforms),
            new SerializedTubeResource<E>(new TubeInterface<E>(tube), platforms)
        );
    }
}
